### PR TITLE
MAINT: logsumexp: properly handle complex sign

### LINKED ISSUE #19854: scipy.special.logsumexp for complex input with return_sign=True under numpy 2.0
### Describe your issue.

https://github.com/numpy/numpy/pull/25441 changes the definition of `np.sign` for complex input in numpy 2.0. When exploring the effect of this change, I found that it leads to strange results for `scipy.special.logsumexp` with `return_sign = True`.

I've demonstrated this below by patching `np.sign` to the new definition. I'm not sure how to interpret the meaning of the `sign` output here for `logsumexp`, but it seems like the change upstream might break some assumptions made by downstream callers of `special.logsumexp`.

Perhaps `scipy` should change how `sign` is computed here, so that it is always $\pm 1$, even for complex input?

### Reproducing Code Example

```python
import numpy as np
from scipy.special import logsumexp

def newsign(x):
  x = np.asarray(x)
  return np.where(x == 0, x, (x / np.abs(x)).astype(x.dtype))

def test():
  rng = np.random.default_rng(0)
  x = rng.normal(size=10) + 1j * rng.normal(size=10)
  print("sign =", np.sign(x.sum()))
  print("log(sum(exp(x))) =", np.log(np.sum(np.exp(x))))
  result, s = logsumexp(x, return_sign=True)
  print("logsumexp result:", result, s)
  print("reconstructed sum(exp(x)):", np.exp(result) * s)
  print("manual sum(exp(x))       :", np.sum(np.exp(x)))
  print()

print("NumPy<2.0 behavior")
test()

_np_sign = np.sign
np.sign = newsign
try:
  print("NumPy>=2.0 behavior")
  test()
finally:
  np.sign = _np_sign
```

### Error message

```shell
NumPy<2.0 behavior
sign = (1+0j)
log(sum(exp(x))) = (2.385269178468705-0.5887388825917717j)
logsumexp result: (2.385269178468705-0.5887388825917715j) (1+0j)
reconstructed sum(exp(x)): (9.033280080568113-6.0317984614767175j)
manual sum(exp(x))       : (9.033280080568113-6.031798461476719j)

NumPy>=2.0 behavior
sign = (0.18452108892318747-0.9828285546027857j)
log(sum(exp(x))) = (2.385269178468705-0.5887388825917717j)
logsumexp result: (2.385269178468705-0.6332187823262332j) (0.9990109323452934-0.044465234223914865j)
reconstructed sum(exp(x)): (8.461679537367544-6.8104861133957915j)
manual sum(exp(x))       : (9.033280080568113-6.031798461476719j)
```

### SciPy/NumPy/Python version and system information

```shell
1.11.4 1.23.5 sys.version_info(major=3, minor=10, micro=12, releaselevel='final', serial=0)
Build Dependencies:
  blas:
    detection method: pkgconfig
    found: true
    include directory: /usr/local/include
    lib directory: /usr/local/lib
    name: openblas
    openblas configuration: USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP= HASWELL MAX_THREADS=2
    pc file directory: /usr/local/lib/pkgconfig
    version: 0.3.21.dev
  lapack:
    detection method: pkgconfig
    found: true
    include directory: /usr/local/include
    lib directory: /usr/local/lib
    name: openblas
    openblas configuration: USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP= HASWELL MAX_THREADS=2
    pc file directory: /usr/local/lib/pkgconfig
    version: 0.3.21.dev
  pybind11:
    detection method: config-tool
    include directory: unknown
    name: pybind11
    version: 2.11.0
Compilers:
  c:
    commands: cc
    linker: ld.bfd
    name: gcc
    version: 10.2.1
  c++:
    commands: c++
    linker: ld.bfd
    name: gcc
    version: 10.2.1
  cython:
    commands: cython
    linker: cython
    name: cython
    version: 0.29.36
  fortran:
    commands: gfortran
    linker: ld.bfd
    name: gcc
    version: 10.2.1
  pythran:
    include directory: ../../tmp/pip-build-env-c6c8ru56/overlay/lib/python3.10/site-packages/pythran
    version: 0.14.0
Machine Information:
  build:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
  cross-compiled: true
  host:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
Python Information:
  path: /opt/python/cp310-cp310/bin/python
  version: '3.10'
```

### PR DESCRIPTION
**Pull request by jakevdp**
Fixes #19854

I opted to have `logsumexp` return a sign that matches the convention of `np.sign` for whichever numpy version is installed.

### CODE REVIEW

**Comment by rgommers** [comment]
Thanks @jakevdp! Seems like a reasonable choice to me. I haven't thought about it as hard as you and @mhvk though. @mhvk would you mind having a look?

File: scipy/special/_logsumexp.py
Comment (mhvk) [inline scipy/special/_logsumexp.py:121]: Since this works also for `floating`, does it make more sense to just use it for all cases?

**Review by mhvk — APPROVED** [review]
Great! In-line a comment about whether we should just use the "correct" sign all the time, but since what you have gives exactly the same answer, either route is fine.

p.s.  Like your numpy-version-agnostic test!

File: scipy/special/_logsumexp.py
Comment (jakevdp) [reply scipy/special/_logsumexp.py:121]: I didn't do that because it makes the implementation more expensive in the case that already works correctly. Would you prefer erring toward simplicitly of implementation?

Comment (jakevdp) [reply scipy/special/_logsumexp.py:121]: If we were opting to always use the numpy 2.0 definition of complex sign, a simpler option would be `s = abs(s)`.

Comment (mhvk) [reply scipy/special/_logsumexp.py:121]: Indeed, `s = abs(s)` would make the intent much clearer.... 

Since a logarithm is taken, I think the division does not matter too much. But a fast alternative would be
```
s *= sgn.conjugate()
```
which would I think be consistent for all cases too. 

But perhaps we're overthinking!?

Comment (jakevdp) [reply scipy/special/_logsumexp.py:121]: Fundamentally, I think the question is whether we want the sign here to be `np.sign(x)`, which changes depending on numpy version, or `x / abs(x)` in all cases.

Comment (mhvk) [reply scipy/special/_logsumexp.py:121]: Ah, yes, it might make sense to just define it usefully for all numpy versions. But I should leave that to the scipy maintainers!

**Comment by jakevdp** [comment]
Updated based on discussion in https://github.com/scipy/scipy/issues/19854#issuecomment-1889543383

File: scipy/special/_logsumexp.py
Comment (mdhaber) [inline scipy/special/_logsumexp.py:46]: While we're at it, isn't this technically about the exponential of the sign of the result, assuming "result" is what the function returns?

File: scipy/special/_logsumexp.py
Comment (mhvk) [inline scipy/special/_logsumexp.py:102]: Had to think this through: the goal of `a_max` is to stop overflow, for which only `a.real.max()` matters. In principle, large imaginary numbers are of course problematic too, since then the argument becomes indeterminate. But not obvious it helps to subtract that from `a` before doing the exponential. So, I *think* this can be `a_max = np.amax(a.real, ...)` unconditionally (with a comment stating we do `.real` for complex perhaps).

Separately: the indentation is off here.

File: scipy/special/_logsumexp.py
Comment (mhvk) [inline scipy/special/_logsumexp.py:122]: I don't know what the scipy convention is here, but in astropy we'd explicitly do something like
```
if NUMPY_LT_2_0 and ...:
```
so that once the minimum supported version of numpy becomes 2.0, we have an easy way of
removing the work-around (since then `np.sign` itself will do the right thing in numpy >= 2.0).

File: scipy/special/_logsumexp.py
Comment (mhvk) [inline scipy/special/_logsumexp.py:126]: Since we're slightly changing API anyway, would it make sense to not do `.astype(s.dtype)`? With the sign taken out, the answer should arguably be float even for complex.

**Review by mhvk — COMMENTED** [review]
looking at this more, one wonders whether this function really makes sense for complex, but some comments in-line about it. Only first one about `a_max` is important.

File: scipy/special/_logsumexp.py
Comment (jakevdp) [reply scipy/special/_logsumexp.py:102]: The reason I changed this is because if `a_max` is complex, it affects the phase and so the sign computation below does not return a useful result.

File: scipy/special/_logsumexp.py
Comment (jakevdp) [reply scipy/special/_logsumexp.py:122]: I couldn't find any examples of this in the library (aside from a couple tests and imports) so I decided not to introduce that kind of logic.

File: scipy/special/_logsumexp.py
Comment (jakevdp) [reply scipy/special/_logsumexp.py:126]: Sure, that's a possiblity. Let's see what the maintainers think.

File: scipy/special/_logsumexp.py
Comment (jakevdp) [reply scipy/special/_logsumexp.py:102]: I think since the magnitude of the exponential is dictated by the real part, and the phase is dictated by the imaginary part, you're right that `a.real` probably makes more sense.

File: scipy/special/_logsumexp.py
Comment (jakevdp) [reply scipy/special/_logsumexp.py]: Done

File: scipy/special/_logsumexp.py
Comment (jakevdp) [reply scipy/special/_logsumexp.py]: I think this is a good suggestion. Done.

**Review by mhvk — COMMENTED** [review]
Here's one more suggested change (I had it as a pending comment for about a week it seems...)

File: scipy/special/_logsumexp.py
Comment (jakevdp) [reply scipy/special/_logsumexp.py]: Done, thanks!

**Review by rgommers — APPROVED** [review]
LGTM, let's give this a go - thanks @jakevdp and @mhvk!

**Pull request merged**
